[WIP] Optional PGP encryption of emails

mateuszmandera opened the pull request:
Preliminary changes for implementing https://github.com/zulip/zulip/issues/11262

At this point, we only makes some initial model changes, implement a skeleton for pgp_sign_and_encrypt (actual signing and encrypting aren't implemented yet and just raise NotImplementedError), and refactor send_email.py. With encryption and signatures disabled, as they should be, the code works as before and tests pass. 

Of course not mergable yet.

mateuszmandera committed:
pgp_email: Make necessary model changes and migrations.

We add a new table UserPGP, for storing users' public keys. We also add
fields want_encrypted_emails and want_signed_emails to UserProfile,
which store user preferences for whether they want emails zulip will
send to them encrypted and/or signed. Finally, we make the corresponding
migrations and add UserPGP to the lists of tables in export.py.

rht commented:
Is the remaining TODO in writing test cases?

mateuszmandera commented:
Test cases and actually implementing encrypting and signing. And then allowing users to set their preferences and upload their public key.

I suppose just adding test cases could make it mergeable, but since all this code does nothing new until the things mentioned above are finished, I don't know if there's any reason to merge.

mateuszmandera commented:
I edited the "Start implementing pgp_sign_and_encrypt" commit to fix a silly mistake (it had `` except UserProfile.DoesNotExist: `` instead of `` except UserPgp.DoesNotExist: `` in line 31) and rebased.

rht commented:
> I suppose just adding test cases could make it mergeable

Merging chunks of the commits earlier helps with the case when the PR becomes stale after changes from other commits induce a merge-conflict. If you add the test cases first instead of complete-the-feature-then-add-test-cases, you will have less problem with rebases.

mateuszmandera commented:
I imagine the tests, for those unimplemented parts should simply assert that NotImplementedError will be raised?

rht commented:
> I imagine the tests, for those unimplemented parts should simply assert that NotImplementedError will be raised?

Yeah, I'd do that.

timabbott commented on zerver/models.py:
```diff
@@ -2570,3 +2574,7 @@ class BotConfigData(models.Model):
+
+class UserPgp(models.Model):
+    user_profile = models.OneToOneField(UserProfile, on_delete=CASCADE)  # type: UserProfile
+    public_key = models.CharField(max_length=1000)  # type: str
```
Lets use `models.TextField` for the key.

timabbott commented on zerver/models.py:
```diff
@@ -2570,3 +2574,7 @@ class BotConfigData(models.Model):
     class Meta(object):
         unique_together = ("bot_profile", "key")
+
+class UserPgp(models.Model):
```
call this `class UserPGP`

timabbott commented on zerver/lib/send_email.py:
```diff
@@ -110,15 +115,24 @@ def send_email(template_prefix: str, to_user_ids: Optional[List[int]]=None,
+    else:
+        emails = build_emails_by_user_ids(template_prefix, to_user_ids=to_user_ids, from_name=from_name,
+                                          from_address=from_address, reply_to_email=reply_to_email,
+                                          language=language, context=context)
```
I don't like this split since it involves copying a bunch of arguments around; can we have the conditional live in `build_email` instead?

timabbott commented:
@mateuszmandera I posted a few comments.  Generally, the direction looks reasonable.  I think we'll need to be careful about the `To/Cc` part for the `to_user_ids` case -- for that one, we want a single email going to all the recipients for the purposes of both the To line and `reply-all` working.  Might need to play games with the envelope sender to make that work with this.  But I think I'd actually be happy with a version that just only supports PGP encryption for the single user ID case first; we only have a single code path initiated from a command-line tool that sends to multiple users anyway.

mateuszmandera committed:
pgp_email: Start implementing pgp_sign_and_encrypt functionality.

We create zerver/lib/pgp.py, with the function pgp_sign_and_encrypt
which, given a to-be-sent email message and a list of user profiles,
will encrypt and/or sign the message, if appropriate based on server
settings and user preferences, and return a list of emails ready to be
sent. We add the option to enable/disable encryption and signatures in
settings.py, by default set to False and should be kept that way until
the features are actually implemented.

At the moment trying to sign or encrypt will cause NotImplementedError
to be raised, but works (basically doing nothing except setting the To
header) in the most basic case where we don't want any
encryption/signatures.

mateuszmandera committed:
pgp_email: Refactor send_email.py to use pgp_sign_and_encrypt.

At this point the original behavior is retained, so with
encryption/signatures settings disabled sending emails works as it did
before and is only refactored and old tests should be passing.

mateuszmandera committed:
email: Move a couple of useful functions to email_helpers.py.

mateuszmandera committed:
pgp_email: Add a preliminary implementation of dev_send_pgp_email command.

We add a management command dev_send_pgp_email. It's meant as a tool in
development for testing sending of emails using PGP features.

mateuszmandera committed:
pgp_email: Add python-gnupg dependency.

We need this for signing/encrypting emails.

mateuszmandera committed:
pgp_email: Use a new PGPEmailMessage class.

We add a PGPEmailMessage class, inheriting from EmailMultiAlternatives,
and switch to using it. We need our own class to have more flexibility
and comfort in what we do with MIME headers.

mateuszmandera committed:
pgp_email: Create a stub of test_pgp.py.

We add a stub of tests for the pgp functionalities. For now they don't
do much as signing and encrypting isn't implemented. We use the word
'subject' in the context of emails in the new file, so it has to be
added to the list of exceptions from the linting rule that prohibits
that word.

mateuszmandera committed:
pgp_email: Begin implementing pgp, by overriding .message().

To have the email correctly formatted according to PGP/MIME (RFC3156),
we need full control over the various MIME parts and headers, that we
don't have through EmailMultiAlternatives. When sending an email,
Django invokes the .message() method of EmailMessage which converts it
and returns an object of type used by the python "email" library.
This is the stage at which we have the desired level of control.
(See django.core.mail.message for details.)
Thus, encryption and signing need to happen by overriding .message().

mateuszmandera committed:
pgp_email: Expand test_pgp.py to cover force_single_message=True.

mateuszmandera committed:
pgp_email: Implement PGP/MIME formatting of emails.

When .send() is invoked for an email message, Django sends the output of
.message() through the apprioprate backend. This .message() output is
essentially a MIME object from the python "email" module. If we're
sending an encrypted/signed email, this object needs to have an
appropriate PGP/MIME structure, as described in RFC3156. In this commit,
we implement the creation of this structure. Instead of actual
ciphertexts/signatures, we insert dummy strings for now, as this last
part of the process isn't implemented yet.

Appropriate tests are added to test_pgp.py.

mateuszmandera commented:
Next stage has been implemented: we move to using a new class - PGPEmailMessage, inheriting from django's EmailMultiAlternatives. When .sign() or .encrypt() methods get invoked, PGP-magic happens and we generate an appropriate MIME structure for an OpenPGP email. Instead of actual signatures/ciphertextx we just insert dummy ones, as this is the next stage and thus not implemented.
Various tests are included.

What's left is to implement using python-gnups to generate the pgp signatures/ciphertexts so they get inserted into the MIMEs instead of the dummy ones. And then frontend stuff.

mateuszmandera committed:
pgp_email: Finish implementing the backend - add gpg encrypt/sign.

We implement actual encryption/signing functions, using the gnupg
module. Whenever one of the gpg functions are run, a new temporary
directory is created and passed as the homedir argument to a new GPG
instance. This ensures resources aren't shared and we don't run into
weird race conditions when the gpg functions are run simultanously by
different threads/processes, as well as ensuring a fast cleanup, as we
can simply delete the entire directory.

The functions take necessary for their operation private/public keys as
argument and import them into their temporary gnupghome to use them to
encrypt/decrypt content. Signature creation/verification functions
import the server key, fetched by the _get_server_gpg_key() function.
The key is supposed to be located at the path in
settings.GPG_SERVER_KEYFILE, which defaults to
zproject/gpg_server_key.asc as of now, as this is meant for the dev
environment only. This should be adjusted to use an apprioprate file
in /etc/zulip in production, when the feature is fully ready.

mateuszmandera committed:
pgp_email: Expand tests.

mateuszmandera committed:
pgp_email: Add support for exporting the UserPGP table.

mateuszmandera committed:
pgp_email: Make tests verify that signed content is as expected.

mateuszmandera commented:
Some comments to perhaps help in review, about decision I made and so on:

1. As explained in a comment earlier, pgp_sign_and_encrypt by default can return a list of emails to be sent. However, it supports force_single_message=True, where it will ensure only one email is generated (with all the recipients in To:), while making the best effort to ensure maximum security without violating any recipient's preferences. If all recipients want encryption, then we can encrypt. If all recipients want signatures then we sign.
A question that I mentioned earlier - should we give up on signing if even a single recipient doesn't have want_signed_emails=True? Now I think the answer is yes, because a signed email has a very particular MIME structure, and for example if the recipient is some kind of bot, it likely will only be able to read basic emails, but won't understand all the PGP/MIME magic. We don't want to break things, so we shouldn't sign.

2. Using PGPEmailMessage class doesn't introduce regressions, because the class behaves same as EmailMultiAlternatives for "vanilla" messages, when .sign() or .encrypt() don't get called. It passes tests from the email testing suite from django's source code (I ran that as an experiment and for assurance).

3. The python-gnupg library is clunky to use, but there's nothing better until we can switch to gpgme python bindings. But that seems difficult until all supported linux distributions have it packaged (Trusty and Xenial only have an ancient, incompatible version that we don't want to use).

4. dev_send_pgp_email is a clunky management command that allows to test and see PGP working in the wild - first email sending via SMTP needs to be configured in the dev environment
```
# SMTP settings for forwarding emails sent in development
# environment to an email account.
EMAIL_HOST = "smtp.gmail.com"
EMAIL_HOST_USER = "xxxxx@gmail.com"
EMAIL_USE_TLS = True
EMAIL_PORT = 587
```
as an example. The command uses the smtp backend by default instead of the dev environment backend, Then a pgp key should be put in the file zproject/gpg_server_key.asc
Example uses are: 
``  ./manage.py dev_send_pgp_email --recipient=recipient@gmail.com --fixture=zerver/tests/fixtures/email/1.txt  --sign ``
to send a signed, unencrypted email.
`` ./manage.py dev_send_pgp_email --recipient=recipient@gmail.com --fixture=zerver/tests/fixtures/email/1.txt  -d --public-key=gpgkey.asc --sign ``

Where gpgkey.asc is a file with the recipient's public key in it.

`` --backend=console `` can be used to  have the email .as_string() printed to stdout instead of sending it out.
The command is ugly and clunky, as for now it's only meant for testing if things work while the feature is being developed.

mateuszmandera commented:
Resolved conflict and cleaned up usage of a highly unnecessary method _cache_original_message() in PGPEmailMessage.
